### ISSUE TITLE
SHAPR for custom defined models

### ISSUE THREAD

**Issue by mehwish2021** [NorskRegnesentral/shapr#420]
Hello All, 
I am using DeepCC model in which I want to apply SHAPR to get the shaply values . 
I defined the functions as described in here https://cran.r-project.org/web/packages/shapr/vignettes/understanding_shapr.html , but its still giving me the error that SHAPR is not defined for this model.

I have defined these functions get_model_specs() and predict_model() as mentioned in the vignette but I am confused how and where are those functions called?
Please help as it has taken much time and I am unable to resolve this issue

**Comment by martinju** [comment]
Hi! Please use the github version of the shapr package.

**Comment by mehwish2021** [comment]
Thanks for your prompt response Martinju.  
A little more clarification, do you mean I have to add these functions in the code of shapR package?? Can you please guide in which file I need to add these functions ?
I would really appreciate you response

**Comment by martinju** [comment]
No, the package on CRAN is outdated (and will be replaced by the main at GitHub) soon. It does things a bit differently. I suggest you install the github version of shapr instead, modify your code accordingly, try to use that to explain your custom model instead. The procedure is explained her:
https://norskregnesentral.github.io/shapr/articles/understanding_shapr.html#explain-custom-models
(which is slightly different from the procedure with the version on CRAN)

If you still got issues after that, let me know.

**Comment by vmombo** [comment]
Hello All, I have a Keras model, is it possible to make it work with it?

**Comment by martinju** [comment]
> Hello All, I have a Keras model, is it possible to make it work with it?
> 

Yes. Please see the main vignetten on the pkgdown site for instructions.

**Comment by martinju** [comment]
Oh, thanks a lot for catching that bug! I will fix it ASAP

**Comment by martinju** [comment]
@vmombo I just merged the fix to main

**Comment by vmombo** [comment]
thank you very much. I have just seen your comment

**Comment by vmombo** [comment]
Hi @martinju , 

I've 85 variables to explain for some predictions, I've tried several times but I got some crashes with the following error : `R(27812,0x2033e3840) malloc: *** error for object 0x600001bdc080: pointer being freed was not allocated
R(27812,0x2033e3840) malloc: *** set a breakpoint in malloc_error_break to debug`, 

I’m not sure if this is due to my computer's capacity, but I suspect it might be (so many coalitions).  I found a "walk around" which is  by grouping variables as you did in the paper.

However and most importantly, I would like to know the following:
- In the grouped approach, what exactly is being explained in the predictions?
- Is it the mean value of the group?

Thanks in advance, for your help

**Comment by martinju** [comment]
Hi again!
85 variables is a lot, so resorting to the grouped approach is recommended. 
By reducing the number of coalitions evaluated in each batch, you should be able to avoid memory issues, however. But it will take a lot of time to compute shapley values with decent accuracy.

The grouped approach explains how the group of features contributes to the prediction instead of the individual features (i.e. what happens as you add/remove the entire group instead of add/remove individual features) -- it is not about mean values or so. A basic and practical introduciton to the approach is given here: https://martinjullum.com/publication/jullum-2021-efficient/jullum-2021-efficient.pdf

**Comment by vmombo** [comment]
Thank you for the clarification and the resource!

I'll take a closer look at the reference—appreciate you sharing it!
